### DISCUSSION TITLE
Deleting from a List using a ConfirmationDialog in a child view.

### DISCUSSION THREAD

**Discussion by tylerjames** [pointfreeco/swift-composable-architecture#1817] [Q&A] [answered]
I have a parent feature that holds an `IdentifiedArrayOf<ChildState>`. It presents a list of `NavigationLink`s using `ForEachStore(ChildState`. The parent reducer also uses `.forEach` to embed child reducers.

The parent feature looks like this:
```swift
struct SavedNetworks: ReducerProtocol {
 struct State: Equatable {
    var networks: IdentifiedArrayOf<NetworkDetails.State>
  }

  enum Action {
    case networkDetailsAction(id: String,
                              action: NetworkDetails.Action)
  }

  var body: some ReducerProtocol<State, Action> {
    Reduce { state, action in
      switch action {  
        case .networkDetailsAction(let id, .delegate(let delegateAction)):
            // Handle delete tapped
          switch delegateAction {
            case .deleteConfirmed:
              state.networks.remove(id: id)
              return .none
          }
      }
    }
    .forEach(\.networks,
              action: /Action.networkDetailsAction(id: action:)) {
      NetworkDetails()._printChanges()
    }
  }
}
```

The user can navigate into a Details screen by tapping on a list item. Inside the Details screen is a **Delete** button. This shows a confirmation dialog that asks the user if they are sure. If they confirm the delete then the action `.delegate(.deleteConfirmed)` is sent and the parent reducer removes the child state from the collection.

**Here's the problem:**  
After the delete action is sent and the child state is removed from parent's collection the confirmation dialog sends a `.dialogDismissed` action sets the confirmation dialog state to `nil` in he child, but it also get picked up by the parent which causes some warnings to be printed to the console:
> • This action was sent to the store while its state contained no element at this ID.

Additionally, if I tap the back button in the navigation bar (because the child view does not get dismissed) and go back to the list view the confirmation dialog is shown again, but this time in the list view.

Should I be handling this confirmation dialog differently? Or maybe I need to force child view to be popped off the stack after it has been deleted?

**Comment by tylerjames** [comment, marked as answer]
Okay. I managed to get it working by handling the confirmation dialog actions differently. When the user selects **Delete** from the confirmation dialog I set its state to `nil` immediately and _then_ send the `.deleteConfirmed` action to the parent reducer.

As for navigation I resorted to doing a manual dismiss of the detail view using a method @mbrandonw described here:
https://forums.swift.org/t/programatic-dismiss-navigation-animation-based-on-state-binding/39275/7

Basically hold an `isPresented` bool in the state and then change that to false when appropriate. Watch for the change in an `.onChange` modifier like this:
```swift
@Environment(\.presentationMode) var presentationMode: Binding<PresentationMode>

.onChange(of: viewStore.isPresented) { isPresented in
  if !isPresented {
    self.presentationMode.wrappedValue.dismiss()
  }
}
```
Beginning to understand why people had some qualms with SwiftUI navigation.

**Reply by mbrandonw** [reply]
Hi @tylerjames, glad you figured it out! We are hopeful that the new navigation APIs we are working on for TCA will soften the rough edges of situations like this. But at least you got something working for now. 🙂

**Reply by tylerjames** [reply]
Looking forward to it! Hopefully it can smooth some things over in iOS 14 as well, even though I'm sure the newer navigation APIs in iOS 16 are better to work with.
